README
NAME
    Debuggit - A fairly simplistic debug statement handler

SYNOPSIS
      use Debuggit(DEBUG => 1);

      # say you have a global hashref for your site configuration
      # (not to imply that global vars are good)
      our $Config = get_global_config();

      # now we can set some config things based on whether we're in debug mode or not
      $Config->{'DB'} = DEBUG ? 'dev' : 'prod';

      # maybe we need to pull our local Perl modules from our VC working copy
      push @INC, $Config->{'vcdir/lib'} if DEBUG;

      # basic debugging output
      debuggit("only print this if debugging is on");
      debuggit(3 => "only print this if debugging is level 3 or higher");

      # show off our formatting
      my $var1 = 6;
      my $var2;
      my $var3 = " leading and trailing spaces   ";
      # assuming debugging is enabled ...
      debuggit("var1 is", $var1);   # var1 is 6
      debuggit("var2 is", $var2);   # var2 is <<undef>>
      debuggit("var3 is", $var3);   # var3 is << leading and trailing spaces   >>
      # note that spaces between args, as well as final newlines, are provided automatically

      # use "functions" in the debugging args list
      my $var4 = { complex => 'hash', with => 'lots', of => 'stuff' };
      # this will call Data::Dumper::Dumper() for you
      # (even if you've never loaded Data::Dumper)
      debuggit("var4 is", DUMP => $var4);

      # make your own function
      Debuggit::add_func(CONFIG => 1, sub { my ($self, $var) = $_; return "$self var $var is $Config->{$var}" });
      # and use it like so
      debuggit(CONFIG => 'DB');     # CONFIG var DB is dev

DESCRIPTION
    You want sophisticated, full-featured, on-demand debugging, and you
    don't want to take it out when you release the code because you might
    need it again later, but you also don't want it to take up any space or
    cause any slowdown of your production application. Sounds impossible?
    Nah. Just use Debuggit.

Quick Start
    To start:

        use strict;
        use warnings;

        use Debuggit;


        my $var = 6;
        debuggit(2 => "var is", $var);      # this does not print
        debuggit(4 => "var is", $var);      # neither does this

    Later ...

        use strict;
        use warnings;

        use Debuggit(DEBUG => 2);


        my $var = 6;
        debuggit(2 => "var is", $var);      # now this prints
        debuggit(4 => "var is", $var);      # but this still doesn't

    That's it. Really. Everything else is just gravy.

A Note on Documentation
    You know, documentation is a double-edged sword. Put too little, and no
    one will use your module because it's poorly documented. Put too much,
    and no one will use your module because it *must* be too complex ... I
    mean, just look at how much documentation it takes to describe it!

    Well, I happen to actually *like* writing documentation, so there's a
    healthy chunk of it here. But let me assure you that this is just
    because I'm being thorough, not because this module is complex. It's
    actually almost ridiculously simple at its core. (In fact, about 80% of
    the module *is* this documentation.) Start with the "Quick Start" above.
    After you get bored with how simple Debuggit is to use like that, and
    want to explore more options, come back and read the rest. Because
    everything thing else is completely optional.

A Note on Version Numbers
    I know it's very fashionable to mark any module you put on CPAN for the
    first time as version 0.000000001, and then take about eight years
    before daring to release a 1.0. However, this module (in earlier
    incarnations) has been used in production code for over 10 years now, so
    I don't really think it makes much sense to call it 0.01. I've chosen to
    start its CPAN life at 2.01, as this is its 3rd major rewrite. If it
    makes you feel better, just subtract 2 from the version number and
    you'll get a fairly accurate idea of its "true" CPAN version. Just be
    aware that it has had a pretty full life outside CPAN as well.

    The Changes file details this life fairly accurately. The version
    numbers are assigned using 20/20 hindsight, and I've filled in a few
    gaps in historical notes, but all the dates and most of the commit
    messages are fully accurate, and thanks to the wonders of version
    control I still have every previous version (save one, which is so
    marked). Not that you care, I'm sure. But it was fun going back through
    the past ten years of Debuggit's history.

EXPORTS
  DEBUG
    "DEBUG" is a constant (similar to "use constant") which holds your
    current debugging level. Because it's implemented using constant
    folding, any conditional based on it will actually be removed during
    compile-time if the debugging level isn't high enough (or turned off
    completely). For instance, this code:

        calculate_complex_stuff(1..10_000) if DEBUG >= 2;

    would disappear entirely if "DEBUG" is set to 0, or to 1. Of course,
    being a constant has its own foibles: you can't interpolate it into
    double-quoted strings, and you can't put it in front of a fat comma. See
    constant for full details.

    You "set" "DEBUG" when you use Debuggit:

        use Debuggit(DEBUG => 2);

    Once it's set, you can't change it. You also can't set it from outside
    the code (like, from an environment variable), but that's a feature I'd
    consider adding if people thought it was useful.

    If you want your debugging turned off altogether, you can do this:

        use Debuggit(DEBUG => 0);

    Or you can do this:

        use Debuggit;

    But there's a subtle difference between those last two. In your
    top-level script, there's actually no difference at all. But in a
    module, not setting "DEBUG" doesn't mean "DEBUG" should be zero; rather
    it means it should "inherit" the value of "DEBUG" from the top-level
    script; or, to look at it another way, the value of "DEBUG" from the
    top-level script "falls through" to whatever modules are included by it.

    The definition of "top-level script" (to Debuggit, anyway) is "first
    'use Debuggit' statement that is executed by the Perl compiler." Which
    means you should always put "use Debuggit" before the "use" statements
    for other modules (or at least other modules of yours).

    So typically what you would do is just put "use Debuggit" at the top of
    all your code until you're ready to debug. When you hit a problem,
    change to "use Debuggit(DEBUG => 3)" (or whatever level you feel is
    appropriate), but *only* in the top-level script. Then you get debugging
    info from all your modules with one simple change. Or, if you just
    *know* the problem is in module X, you can enable debugging in that
    module only. Convenient, right?

    This may not work as well as you'd like if you can't figure out *where*
    your top-level script actually is (one great example of that is a
    mod_perl environment). But you can still enable the debugging in each
    module, so it isn't tragic. Just not as convenient.

FUNCTIONS
  debuggit([level =>] arg[, ...])
    debuggit() is the only exported function of Debuggit (except for the
    constant, which is technically a function, but you know what I mean).
    When DEBUG is set to 0, you are guaranteed that "debuggit" is an empty
    function. When DEBUG is non-zero, you are guaranteed that "debuggit"
    will do the equivalent of this:

        sub debuggit
        {
            my $level = $_[0] =~ /^\d+$/ ? shift : 1;
            if (DEBUG >= $level)
            {
                @_ = process_funcs(@_);
                my $msg = $formatter->(@_);
                $output->($msg);
            }
        }

    meaning that you can override both $formatter and $output, and you can
    add to or subtract from the functions handled by process_funcs().

  [local] $Debuggit::formatter = coderef
    The $formatter variable allows you to override Debuggit's internal
    format function (see above). For instance, something like this:

        $Debuggit::formatter = sub { shift . "\n" . join('', map { "\t$_\n" } @_) };

        my @list = qw< fred sue joe charlie >;
        debuggit("Names:", @list);

    outputs something like this:

        Names:
            fred
            sue
            joe
            charlie

    although that seems less useful than the default formatter, in general.
    Or maybe you don't like how Debuggit provides spaces and newlines for
    you:

        $Debuggit::formatter = sub { join('', @_) };

    But don't forget that now you've also lost the special handling of
    "undef" and strings with leading or trailing spaces.

    Happily, since $formatter is a variable, you can use "local" to restore
    the previous value at the end of the enclosing block. Better yet, you
    can take advantage of the fact that the default formatter is stored as
    Debuggit::default_formatter() to do clever things like this:

        # add timestamp to debugging (at least for this function/module/whatever)
        local $Debuggit::formatter = sub
        {
            return scalar(localtime) . ': ' . Debuggit::default_formatter(@_);
        };

    or even this:

        # all debugging statements in the current context will show function name
        local $Debuggit::formatter = sub
        {
            # note that caller(0) would be this formatter sub, and
            # caller(1) would be debuggit(), so caller(2) is what we want
            # element 3 is the subroutine name (which includes the package name)
            return (caller(2))[3] . ': ' . Debuggit::default_formatter(@_);
        };

    Note that that last example only handles the simple cases--if your
    debuggit() calls get stuck inside eval's or coderef's or anything like
    that, this breaks down. But often the simple case is close enough.

    Functions are handled before the formatter is called (see "DEBUGGING
    FUNCTIONS"), so any replacement formatter you create doesn't have to
    worry about those.

    The default_formatter() looks *mostly* like this:

        $default_formatter = sub
        {
            return join(' ', map { defined $_ ? $_ : '<<undef>>' } @_) . "\n";
        };

    but see below for full details.

    Before you try to get too terribly fancy with the formatter, you may
    wish to investigate add_func(). A good rule of thumb when trying to
    decide whether you want a new function or a new formatter is this:
    imagine that debuggit() takes a bunch of chunks of text and produces a
    line (which is pretty much what is does). If you want to fiddle with how
    one of those chunks looks, you want a function. If you need to change
    the look of the whole line, though, you need a new formatter.

    If you're more interested in where the formatted output gets sent to,
    look at $output.

  Debuggit::default_formatter(@_)
    This function is what $formatter (see above) is set to unless (and
    until) you change it. It can also be called from your own formatter
    function (again, see above). Its purpose is to turn the arguments you
    pass "debuggit()" into a formatted line. This line is then sent to the
    function stored in $output.

    The default formatter provides the following conveniences:

    *   A single space is put between separate arguments.

    *   An undefined argument is replaced with the string '<<undef>>'
        (distinguishes undef from an empty string, and avoids unsightly
        "uninitialized variable" warnings).

    *   Any value which has leading or trailing spaces (that is, / +/, not
        /\s+/) has '<<' prepended to it and '>>' appended to it. This allows
        you to easily see (and hopefully accurately count) any such extra
        spaces.

    *   A newline is appended to the formatted line.

    All these features are demonstrated in the "SYNOPSIS".

  [local] $Debuggit::output = coderef
    The $output variable allows you to override Debuggit's internal output
    function (see debuggit()). For instance, something like this:

        local $Debuggit::output = sub { print @_ };

    allows you to print debugging messages to stdout rather than stderr
    (although I'm not sure why you'd want to). Like with $formatter, the use
    of "local" allows you to change the output function temporarily (i.e.
    until the end of the enclosing block).

    Perhaps you want a log file:

        my $log = '/tmp/debug.log';
        $Debuggit::output = sub
        {
            open(LOG, ">>$log") or return;
            print LOG @_;
            close(LOG);
        };

    (Notice how you have to append to the file, else multiple debuggit()
    calls will just overwrite each other.)

    Note that you don't have to append a newline ($formatter does that). And
    finally note that *not* using "local" sometimes has its advantages: in
    this case, you might put such code in a common header file that all your
    Perl modules call, and the output will be adjusted for all parts of your
    program, regardless of scope. (See "POLICY MODULES" for the best way to
    accomplish that.)

    You could also save to a string:

        our $log_msg;
        local $Debuggit::output = sub { $log_msg .= join('', @_) };

    Again, we're appending. We join all the args together (although most
    formatters will return only one value, probably best not to assume), but
    use no separator. Also note the use of "our"; you probably want that
    rather than a "my" variable.

    The default output function is merely:

        sub { print STDERR @_ };


Changes
Revision history for Perl module Debuggit


2.07 Apr  9, 2018
    - update to work with Data::Printer 0.36 and above
	- lots of misc POD tweaks
    - distro fixes:
        - make it so an old version of Data::Printer fails a unit test
          (instead of generating a FAIL report from CPAN Testers)

2.06 Mar 31, 2015
    - fix warnings in newer versions of Perl
        - thanks to ANDK for the patch
        - closes RT/90433
    - test suite stuff:
        - rename nomem1.t to nomem-old.t
        - do a skip_all in all cases
            - this avoids GTop failures on certain Perl versions
        - rename nomem2.t to nomem.t
        - Memory::Usage is now the official way to test we're not adding RAM

2.05 Nov  5, 2013
    - more distro fixes:
        - remove evil stuff that had crept into MANIFEST
        - updated MANIFEST.SKIP to hopefully keep that from happening again
        - moved POD tests to xt/ so they can't cause install failures
        - fixed Makefile.PL to generate both META.yml *and* META.json

2.04 Feb 15, 2013
    - added support for Data::Printer via a new config option
    - DUMP function (with Data::Dumper) now sorts hash keys
    - doco for DataPrinter option, plus more doco tweaks
    - fixed many distribution/testing errors:
        - added configure_requires for YAML::Tiny
            - should fix UNKNOWN reports on CPAN Testers
        - revamped default_funcs.t to work on Windows
        - added testing Perl snippet framework to work across Windows/Linux
            - working with DANBOO to get this stuff into Test::Command
        - now sorting Data::Dumper hash keys in all tests too

2.03 Dec 30, 2012
    - very slight doco tweaks
    - fixed Module::Install version as per Miyagawa
        (http://weblog.bulknews.net/post/33907905561/do-not-ship-modules-with-module-install-1-04)

2.02 Feb 19, 2012
    - fixed annoying "redefined" warnings
    - fixed annoying "naked subroutines" test failures for callers of Debuggit
        - proper handling of Pod::Coverage's absence in t/pod_coverage.t
    - can now override the system wide DEBUG for a specific module
    - adding another test to insure no additional memory is used
        - refactored t/nomem1.t to match t/nomem2.t
        - more cautious handling of Memory::Usage to avoid non-Linux problems
    - updated / extended all POD
        - now have main POD, manual, and cookbook
    - fixed the problem with Moose classes autocleaning DEBUG

2.01 Jul 29, 2011
    - initial CPAN release
    - updates to POD
    - fixed Changes file (now in reverse chronological order)

2.00 Mar 29, 2011
    - initial version as Debuggit.pm
        - rebuilt from the ground up (again)
        - still using TDD; now using TAP
        - all previous features retained except:
            - messing about with VC working dirs removed
            - that probably should have always been separate, really
    - designed to be released onto CPAN
        - distro built up (using ExtUtils::ModuleMaker::PBP)
        - added POD
        - tests fully TAP compliant
        - added lots of comparisons with other CPAN modules
        - now verifying claims of 0 memory and 0 code via tests
    - added some new features:
        - removed dependence on Exporter (wasn't really using it for anything anyway)
        - DEBUG moved to package Debuggit:: (of course)
        - DEBUG already defined in caller not an exception (similar to VCtools::Base)
            - can't remember why I thought it should be an error in the first place
        - min debugging level (first arg to debuggit) is now optional (default: 1)
        - separated out formatter so you can change how the debugging output looks
        - separated out output method so you can change where the debugging output goes
        - exposed default_formatter as part of the public interface
        - formalized the debugging functions
            - added $self arg
            - added add_func
            - added remove_func
        - added POLICY arg to import for exporting up a level
    - tested all this in production code for roughly two years before deemed ready for CPAN
        - as noted below, original Debuggit.pm dates from Aug 7, 2008


[END OF LIFE (such as it was) for Company::Debug]

1.99 Summer, 2008
    - initial version as Company::Debug
    - first idea for debugging functions
        - default function DUMP for runtime require'ing of Data::Dumper
        - flexible multi-arg functions for handling repititious debugging like in the POOF code
            - so, thanks to BMILLARES for giving me the brainstorm! (albeit inadvertently)
            - the OBJDATA example in Debuggit.pm is almost verbatim what Benny was doing
    - eventually abandoned this idea in favor getting Debuggit.pm ready for CPAN

[creation of Company::Debug]
[this idea was never completed, and consequently never checked in]


[END OF LIFE for Barefoot.pm]

1.13 Jul 11, 2009
    - cleaned up super-obscure bug caused by trying to call debuggit() from inside _set_debuggit_func()
        - not only avoids spurious warning, but should provide better performance as well

1.12 Jul 11, 2009
    - fixed bug where testing modules redirect didn't work when Debuggit was loaded
    - trying new, simpler method of redirecting to testing modules
        - not sure why I was doing it so complicated-ly anyway ...
    - fixed some debugging statements and added a few more

1.11 Jul 9, 2009
    - think I managed to fix the Debuggit/Barefoot conflicts, finally

1.10 Jun 24, 2009
    - now allows Debuggit.pm to override DEBUG and debuggit(), as long as it's loaded first
        - any DEBUG parameter to Barefoot is ignored
        - Barefoot modules are still redirected to test areas
    - fixed bug where call to croak() bombs because Carp wasn't loaded
      (this bug was hardly ever hit in actual code)


[work on 2.x branch (actual Debuggit.pm) started here: Aug 7, 2008]

1.04 Jun 23, 2008

lib/Debuggit.pm
package Debuggit;

use strict;
use warnings;

our $VERSION = '2.07';


#################### main pod documentation begin ###################
####
###
##
#

=head1 NAME

Debuggit - A fairly simplistic debug statement handler

=head1 SYNOPSIS

    use Debuggit DEBUG => 1;

    # say you have a global hashref for your site configuration
    # (not to imply that global vars are good)
    our $Config = get_global_config();

    # now we can set some config things based on whether we're in debug mode or not
    $Config->{'DB'} = DEBUG ? 'dev' : 'prod';

    # maybe we need to pull our local Perl modules from our VC working copy
    push @INC, $Config->{'vcdir/lib'} if DEBUG;

    # basic debugging output
    debuggit("only print this if debugging is on");
    debuggit(3 => "only print this if debugging is level 3 or higher");

    # show off our formatting
    my $var1 = 6;
    my $var2;
    my $var3 = " leading and trailing spaces   ";
    # assuming debugging is enabled ...
    debuggit("var1 is", $var1);   # var1 is 6
    debuggit("var2 is", $var2);   # var2 is <<undef>>
    debuggit("var3 is", $var3);   # var3 is << leading and trailing spaces   >>
    # note that spaces between args, as well as final newlines, are provided automatically

    # use "functions" in the debugging args list
    my $var4 = { complex => 'hash', with => 'lots', of => 'stuff' };
    # this will call Data::Dumper::Dumper() for you
    # (even if you've never loaded Data::Dumper)
    debuggit("var4 is", DUMP => $var4);

    # or maybe you prefer Data::Printer instead?
    use Debuggit DEBUG => 1, DataPrinter => 1;
    debuggit("var4 is", DUMP => $var4);

    # make your own function
    Debuggit::add_func(CONFIG => 1,
            sub { my ($self, $var) = $_; return (lc($self), 'var', $var, 'is', $Config->{$var}) });
    # and use it like so
    debuggit(CONFIG => 'DB');     # config var DB is dev


=head1 DESCRIPTION

You want debugging?  No, you want sophisticated, full-featured, on-demand debugging, and you don't
want to take it out when you release the code because you might need it again later, but you also
don't want it to take up any space or cause any slowdown of your production application.  Sound
impossible?  Nah.  Just use Debuggit.


=head2 Quick Start

To start:

    use strict;
    use warnings;

    use Debuggit;


    my $var = 6;
    debuggit(2 => "var is", $var);      # this does not print
    debuggit(4 => "var is", $var);      # neither does this

Later ...

    use strict;
    use warnings;

    use Debuggit DEBUG => 2;


    my $var = 6;
    debuggit(2 => "var is", $var);      # now this prints
    debuggit(4 => "var is", $var);      # but this still doesn't

That's it.  Really.  Everything else is just gravy.


=head2 Documentation

This POD explains just the basics of using C<Debuggit>.  For full details, see L<Debuggit::Manual>.

=cut

#
##
###
####
#################### main pod documentation end ###################

my ($debuggit, $add_func);


#####################################################################
##
#

=head1 EXPORTS

=head2 DEBUG

DEBUG is a constant integer set to whatever value you choose:

    use Debuggit DEBUG => 2;

or to 0 if you don't choose:

    use Debuggit;

Actually, failure to specify a value only defaults to 0 the first time in a program this is seen.
Subsequent times (e.g. in modules included by the main script), DEBUG will be set to the first value
passed in.  In this way, you can set DEBUG in the main script and have it "fall through" to all
included modules.  See L<Debuggit::Manual/"The DEBUG Constant"> for full details.

=head2 Functions exported

The L</debuggit> function is exported. The Alias parameter can be used to provide a shorter alias for calling this function, for example:

    use Debuggit Alias => 'dbg';  # The dbg() function is now an alias to debuggit()
    # And then:
    dbg("This message will be printed");

The user is warned that the alias that they choose might conflict with other modules. More warning and advice are given in the L<Debuggit::Manual/"Creating an alias for the debuggit() function"> page.

=cut

#
##
#####################################################################


sub import
{
    my ($pkg, %opts) = @_;
    my $caller_package = $opts{PolicyModule} ? caller(1) : caller;

    my $master_debug = eval "Debuggit::DEBUG()";
    my $debug_value = defined $opts{DEBUG} ? $opts{DEBUG} : defined $master_debug ? $master_debug : 0;
    unless (defined $master_debug)
    {
        # Perl does not know whether the string eval below will modify
        # $debug_value, so it assumes the worst.  So make the constant
        # out of a new lexical scalar outside the eval's visible scope.
        # This quiets a new warning in 5.20.  Thanks ANDK!
        my $inner_val = $debug_value;
        *Debuggit::DEBUG = sub () { $inner_val };
        $master_debug = $debug_value;
    }

    no strict 'refs';
    no warnings 'redefine';

    my $caller_value = eval "${caller_package}::DEBUG()";
    if (defined $caller_value)
    {
        warn("Cannot redefine DEBUG; original value of $caller_value is used") if $debug_value ne $caller_value;
    }
    else
    {
        # Thanx to tye from perlmonks for this line of code, which solves the Pod::Coverage issue
        # (see t/pod_coverage.t).               http://www.perlmonks.org/?node_id=951831
        my $inner_val = $debug_value; # See comment above about $inner_val.
        *{ join('::', $caller_package, 'DEBUG') } = sub () { $inner_val };
    }

    if ($debug_value)
    {
        _setup_funcs($master_debug, $debug_value, $caller_package, $opts{DataPrinter});
    }
    else
    {
        *{ join('::', $caller_package, 'debuggit') } = sub {};
        *Debuggit::add_func = sub {} unless Debuggit->can('add_func');
    }

    # User-defined Alias for the debuggit() function
    if ( my $debuggit_alias = $opts{Alias} ) { 
        *{ join('::', $caller_package, $debuggit_alias) } 
            = sub { goto &{ join('::', $caller_package, 'debuggit') } } 
    }

}   # End import()


sub _setup_funcs
{
    my ($master_debug, $debug_value, $caller_package, $data_printer) = @_;

    no strict 'refs';
    no warnings 'redefine';

    # If our debug value is the same as the master debug value, we're just going to export our own
    # debuggit() function out to the calling package.  In this way, we avoid unnecessary code
    # duplication by every package having its own copy of debuggit().  However, if the two values
    # _don't_ match, it means that we're doing an override, and that in turns means that we _have_
    # to give the calling package its own copy.  This is because debuggit() is actually a closure,
    # with the debug value stored in it.  If we have two different debug values (one for the program
    # as a whole, and a different one for this particular package), we have to have two different
    # debuggit() calls as well.
    if ($debug_value == $master_debug)
    {
        *Debuggit::debuggit = eval $debuggit unless Debuggit->can('debuggit');
        *{ join('::', $caller_package, 'debuggit') } = \&debuggit;
    }
    else
    {
        *{ join('::', $caller_package, 'debuggit') } = eval $debuggit;
    }

    unless (Debuggit->can('add_func'))
    {
        eval $add_func;

        # create default function
        if ($data_printer)
        {
            add_func(DUMP => 1, sub
            {
                require Data::Printer;
                Data::Printer->VERSION("0.36");
                shift;
                return &Data::Printer::np(shift, colored => 1, hash_separator => ' => ', print_escapes => 1);
            });
        }
        else
        {
            add_func(DUMP => 1, sub
            {
                require Data::Dumper;
                shift;
                local $Data::Dumper::Sortkeys = 1;
                return Data::Dumper::Dumper(shift);
            });
        }
    }
}


#####################################################################
##
#

=head1 FUNCTIONS

=cut

#####################################################################
##
#

=head2 debuggit

Use this function to conditionally print debugging output.  If the first argument is a positive
integer, the output is printed only if DEBUG is set to that number or higher.  If the first argument
is I<not> a positive integer, the output is printed if DEBUG is non-zero (so omitting the debugging
leve is the same as setting it to 1).  The remaining arguments are concatenated with spaces, a
newline is appended, and the results are printed to STDERR.  Some minor formatting is done to help
distinguish C<undef> values and values with leading or trailing spaces.  To get further details, or
to learn how to override any of those things, see L<Debuggit::Manual/"The debuggit function">.

=head2 default_formatter

This is what C<debuggit> is set to initially.  You can call it directly if you want to "wrap"
C<debuggit>.  For examples of this, see L<Debuggit::Cookbook/"Wrapping the debugging output">.

=cut

#
##
#####################################################################

BEGIN
{
    # This is an anonymous closure.  It has to be both of those things.
    #   *   It has to be anonymous because it may be put into different packages depending on the
    #       circumstances.  See the comments in _setup_funcs() for further details on that.
    #   *   It has to be a closure because we want the debug value (against which we have to check
    #       the first arg, if it's a positive integer), to be stored with the sub.  We in turn want
    #       this for several reasons:
    #       -   We have to reference the DEBUG value in the calling package.
    #       -   If we determine that via reference, that works for most cases.  But in the case of
    #           Moose classes, most of which are autocleaned, the DEBUG constant, which is just a
    #           function, may well be gone by the time debuggit() runs.  If we were calling it
    #           directly, autocleaning wouldn't keep that from working.  But calling by reference is
    #           a whole different story.
    #       -   So our best bet is to use a closure.  The $debug_value referred to below must exist
    #           in the scope in which this is eval'ed.  Then that value gets wrapped in the closure
    #           and it doesn't matter a whit if the function is autocleaned.
    $debuggit = q{
        sub
        {
            return unless @_ > 0 && ($_[0] =~ /^\d+$/ ? shift : 1) <= $debug_value;
            $Debuggit::output->($Debuggit::formatter->(Debuggit::_process_funcs(@_)));
        }
    };
}


sub default_formatter
{
    return join(' ', map { !defined $_ ? '<<undef>>' : /^ +/ || / +$/ ? "<<$_>>" : $_ } @_) . "\n";
}

our $formatter = \&default_formatter;

our $output = sub { print STDERR @_ };


#####################################################################
###
##
#

=head2 add_func

=head2 remove_func

Add or remove debugging functions.  Please see L<Debuggit::Manual/"Debugging Functions">.

=cut

#
##
###
#####################################################################


my %PROCS;

BEGIN
{
    $add_func = q{
        sub Debuggit::add_func
        {
            my ($name, $argc, $code) = @_;

            $Debuggit::PROCS{$name} = { argc => $argc, code => $code };

            return 1;
        }
    };
}


sub remove_func
{
    delete $Debuggit::PROCS{shift()};
    return 1;
}



#####################################################################
# PRIVATE FUNCTIONS
#####################################################################


sub _process_funcs
{
    my @parts;

    while (@_)
    {
        local $_ = shift;

        if ($_ and exists $Debuggit::PROCS{$_})
        {
            my @args = ($_);
            push @args, shift foreach 1..$Debuggit::PROCS{$_}->{argc};
            push @parts, $Debuggit::PROCS{$_}->{code}->(@args);
        }
        else
        {
            push @parts, $_;
        }
    }

    return @parts;
}


#################### remainder of pod begin ###################
####
###
##
#

=head1 DIAGNOSTICS

=over 4

=item * Cannot redefine DEBUG; original value of %s is used

It means you did something like this:

    use Debuggit DEBUG => 2;
    use Debuggit DEBUG => 3;

only probably not nearly so obvious.  Debuggit tries to be very tolerant of multiple imports into
the same package, but the C<DEBUG> symbol is a constant function and can't be redefined without
engendering severe wonkiness, so Debuggit won't do it.  As long as you pass the same value for
C<DEBUG>, that's okay.  But if the second (or more) value is different from the first, then you will
get the original value regardless.  At least this way you'll be forewarned.

=back



=head1 PERFORMANCE

Debuggit is designed to be left in your code, even when running in production environments.
Because of this, it needs to disappear entirely when debugging is turned off.  It can achieve this
unlikely goal via the use of a compile-time constant.  Please see
L<Debuggit::Manual/"Performance Considerations"> for full details.



=head1 BUGS and CAVEATS

=over

=item *

Once you set C<DEBUG>, you can't change it.  Even if you try, you get the original value.  See
L</DIAGNOSTICS>.

=item *

Doing:

    debuggit(0 => "in production mode");

never prints anything, even when C<DEBUG> is 0.  That's because C<debuggit> is guaranteed to be an
empty function when debugging is turned off.

=item *

Doing:

    debuggit($var, "is the value");

is inherently dangerous.  If C<$var> is a positive integer, C<debuggit> would interpret it as a
debug level, and not print it.  So, either do this:

    debuggit(1 => $var, "is the value");

or this:

    debuggit("the value is", $var);

Or, to look at it another way, you can pass a value as the first arg to print, or you can leave off
a debugging level altogether, but don't try to do both at once.

=item *

Doing:

    my $var1 = "DUMP";
    my $var2 = "stuff";
    debuggit(1 => "vars are", $var1, $var2);

is equivalent to:

    debuggit(1 => "vars are", DUMP => $var2);

which is probably not what you wanted, assuming the default functions are still in place.  See
L<Debuggit::Manual/"IMPORTANT CAVEAT!"> for full details.

=item *

Doing:

    debuggit(2 => "first thousand elements:", @array[0..999]);

is likely going to have a performance impact even when debugging is off.  Instead, do:

    debuggit("first thousand elements:", @array[0..999]) if DEBUG >= 2;

See L<Debuggit::Manual/"Style Considerations"> for another example and details on the problem.

=back

That's all I know of.  However, lacking omniscience, I welcome bug reports.



=head1 SUPPORT

Debuggit is on GitHub at barefootcoder/debuggit.  Feel free to fork and submit patches.  Please note
that I develop via TDD (Test-Driven Development), so a patch that includes a failing test is much
more likely to get accepted (or at least likely to get accepted more quickly).

If you just want to report a problem or request a feature, that's okay too.  You can create an issue
on GitHub, or a bug in CPAN's RT (at http://rt.cpan.org).  Or just send an email to
bug-Debuggit@rt.cpan.org.



=head1 AUTHOR

    Buddy Burden
    CPAN ID: BAREFOOT
    Barefoot Software
    barefootcoder@gmail.com

=head1 COPYRIGHT

This program is free software licensed under

    The Artistic License

The full text of the license can be found in the LICENSE file included with this module.


This module is copyright (c) 2008-2015, Barefoot Software.  It has many venerable ancestors (some
more direct than others), including but not limited to:

=over

=item *

C<Barefoot::debug>, (c) 2000-2006 Barefoot Software, 2004-2006 ThinkGeek

=item *

C<Barefoot::base>, (c) 2001-2006 Barefoot Software

=item *

C<Geek::Dev::Debug>, (c) 2004 ThinkGeek

=item *

C<VCtools::Base>, (c) 2004-2008 Barefoot Software, 2004 ThinkGeek

=item *

C<Barefoot>, (c) 2006-2009 Barefoot Software

=item *

C<Company::Debug>, (c) 2008 Rent.com

=back


=head1 SEE ALSO

L<Log::Log4perl>, L<debug>, L<Debug>, L<Debug::Message>, L<Debug::EchoMessage>.

Comparison with most of these (and others) can be found in L<Debuggit::Manual/"Comparison Matrix">.

=cut

#
##
###
####
#################### remainder of pod end ###################


# Return a true value
1;
